crates/uv-workspace/src/pyproject_mut.rs

```diff
@@ -503,13 +503,38 @@ pub fn add_dependency(
     deps: &mut Array,
     has_source: bool,
 ) -> Result<ArrayEdit, Error> {
-    // Find matching dependencies.
     let mut to_replace = find_dependencies(&req.name, Some(&req.marker), deps);
-    match to_replace.as_slice() {
+
+    let edit = match to_replace.as_slice() {
         [] => {
+            // determine the dependency list is sorted prior to
+            // adding the new dependency; the new dependency list
+            // will be sorted only when the original list is sorted
+            // so that users' custom dependency ordering is preserved.
+            let sorted = deps
+                .clone()
+                .into_iter()
+                .collect::<Vec<_>>()
+                .windows(2)
+                .all(|w| w[0].to_string() <= w[1].to_string());
+
             deps.push(req.to_string());
             reformat_array_multiline(deps);
-            Ok(ArrayEdit::Add(deps.len() - 1))
+
+            if sorted {
+                deps.sort_by_key(|d| {
+                    d.as_str()
+                        .and_then(try_parse_requirement)
+                        .unwrap()
+                        .to_string()
+                });
+            }
+            let index = deps
+                .iter()
+                .map(|d| d.as_str().and_then(try_parse_requirement).unwrap())
+                .position(|r| r.name == req.name && r.extras == req.extras)
```

Comment:
not 100% sure if it is correct to omit the comparison for `marker` here

Reply:
If the deps aren't sorted, we should just insert at the end